File: assets/Scripts/player/PlayerUI.js
"use strict";

/**
 * @param {*} v
 * @param {number} fb
 * @returns {number}
 */
function num(v, fb) {
    v = +v;
    return Number.isFinite(v) ? v : fb;
}

/**
 * @param {*} v
 * @returns {boolean}
 */
function isObj(v) {
    return !!v && typeof v === "object" && !Array.isArray(v);
}

/**
 * Fast 2-decimal formatter for quantized integer (value * 100).
 * Produces stable output with minimal work, only when value changed.
 *
 * @param {number} q integer = round(value * 100)
 * @returns {string}
 */
function fmt2FromQ(q) {
    // q can be negative
    const neg = q < 0;
    if (neg) q = -q;

    const i = (q / 100) | 0;
    const f = q - i * 100;

    // Avoid arrays; simple concatenation is fine (called only on change).
    return (neg ? "-" : "") + i + "." + (f < 10 ? "0" : "") + f;
}

/**
 * Deterministic debug HUD for a player.
 *
 * Lifecycle contract:
 * - create(): allocate layer + all elements exactly once
 * - update(model): mutate values only (setText/Visible/Value), no element creation
 * - destroy(): destroy layer deterministically
 */
class PlayerUI {
    /**
     * @param {*} player engine-side player object
     */
    constructor(player) {
        this.player = player;

        const c = (player && player.cfg && player.cfg.ui) ? player.cfg.ui : Object.create(null);

        this.layerName = String(c.layerName || "debug-ui");
        this.anchor = String(c.anchor || "tl");

        this.mx = num(c.marginLeft, 10);
        this.my = num(c.marginTop, 10);

        this.w = num(c.w, 280);
        this.padX = num(c.padX, 12);
        this.padY = num(c.padY, 8);

        this.fontTitle = num(c.fontTitle, 16);
        this.fontLine = num(c.fontLine, 14);
        this.gap = num(c.lineGap, 4);

        this.idPrefix = String(c.idPrefix || "player.debug");

        this.style = isObj(c.style) ? c.style : {
            bg: {color: "#0b0f14", alpha: 0.65},
            border: {size: 1, color: "#8AA0B6", alpha: 0.45, radius: 8}
        };

        this.layer = null;

        // Prebuilt stable ids (no string concat in update()).
        this._id = Object.freeze({
            panel: this.idPrefix + ".panel",
            title: this.idPrefix + ".title",
            fps: this.idPrefix + ".fps",
            pos: this.idPrefix + ".pos",
            camType: this.idPrefix + ".camType",
            camYaw: this.idPrefix + ".camYaw",
            camPitch: this.idPrefix + ".camPitch",
            worldTime: this.idPrefix + ".worldTime",
            timeRate: this.idPrefix + ".timeRate"
        });

        // Dirty-cache (numbers) + last rendered strings.
        this._cache = {
            fps: -1,

            posXq: 0x7fffffff,
            posYq: 0x7fffffff,
            posZq: 0x7fffffff,
            posStr: "",

            camType: "",
            camYaw: NaN,
            camPitch: NaN,

            worldTime: NaN,
            timeRate: NaN
        };
    }

    /**
     * Allocate and build UI once.
     * @returns {PlayerUI}
     */
    create() {
        if (this.layer) return this;

        const HUD = this.player && this.player.d && this.player.d.hud;
        if (!HUD) return this;

        const layer = HUD.layer(this.layerName);
        this.layer = layer;

        // Build once via spec; runtime updates are only setText/setValue/setVisible.
        layer.spec(this._buildSpec(), {relayout: true});

        // Force first full update.
        this._invalidateCache();

        this.update();
        return this;
    }

    /**
     * Pure declarative spec (create-time only).
     * @returns {object}
     */
    _buildSpec() {
        const id = this._id;

        return {
            type: "Panel",
            id: id.panel,
            w: this.w,
            h: 1,
            autoHeight: true,
            padX: this.padX,
            padY: this.padY,
            flow: {
                fontSize: this.fontLine,
                gap: this.gap
            },
            place: {
                anchor: this.anchor,
                x: this.mx,
                y: this.my
            },
            style: this.style,
            children: [
                {
                    type: "Text",
                    id: id.title,
                    text: "DEBUG",
                    fontSize: this.fontTitle,
                    color: "#FFFFFF"
                },
                {type: "Text", id: id.fps, text: "FPS: --"},
                {type: "Text", id: id.pos, text: "POS: --"},
                {type: "Text", id: id.camType, text: "CAM(type): --"},
                {type: "Text", id: id.camYaw, text: "CAM(yaw): --"},
                {type: "Text", id: id.camPitch, text: "CAM(pitch): --"},
                {type: "Text", id: id.worldTime, text: "WorldTime: --"},
                {type: "Text", id: id.timeRate, text: "timeRate: --"}
            ]
        };
    }

    /**
     * Backward-compatible alias (old call sites can keep using refresh()).
     */
    refresh() {
        this.update();
    }

    /**
     * Update values only. No element creation.
     * Uses dirty-check to minimize string allocations and Java calls.
     *
     * @param {*} model optional external model; if not provided uses player.frame
     */
    update(model) {
        const layer = this.layer;
        if (!layer) return;

        const player = this.player;
        const frame = model || (player && player.frame);
        if (!frame) return;

        const pose = frame.pose;
        const view = frame.view;

        // FPS
        const eng = player && player.d && player.d.engine;
        const fps = (eng && typeof eng.fps === "function") ? (eng.fps() | 0) : 0;
        if (fps !== this._cache.fps) {
            this._cache.fps = fps;
            layer.setText(this._id.fps, "FPS: " + fps);
        }

        // Position (quantize *100 and update only on change)
        if (pose) {
            const xq = Math.round(pose.x * 100);
            const yq = Math.round(pose.y * 100);
            const zq = Math.round(pose.z * 100);

            if (xq !== this._cache.posXq || yq !== this._cache.posYq || zq !== this._cache.posZq) {
                this._cache.posXq = xq;
                this._cache.posYq = yq;
                this._cache.posZq = zq;

                const s = "POS: " + fmt2FromQ(xq) + " | " + fmt2FromQ(yq) + " | " + fmt2FromQ(zq);
                this._cache.posStr = s;
                layer.setText(this._id.pos, s);
            }
        }

        // Camera type / yaw / pitch
        if (view) {
            const camType = String(view.type);
            if (camType !== this._cache.camType) {
                this._cache.camType = camType;
                layer.setText(this._id.camType, "CAM(type): " + camType);
            }

            const yaw = +view.yaw;
            if (Number.isFinite(yaw) && yaw !== this._cache.camYaw) {
                this._cache.camYaw = yaw;
                layer.setText(this._id.camYaw, "CAM(yaw): " + yaw);
            }

            const pitch = +view.pitch;
            if (Number.isFinite(pitch) && pitch !== this._cache.camPitch) {
                this._cache.camPitch = pitch;
                layer.setText(this._id.camPitch, "CAM(pitch): " + pitch);
            }
        }

        // World time (guard if WORLD is absent)
        if (typeof WORLD !== "undefined" && WORLD && typeof WORLD.getWorldTime === "function") {
            const wt = WORLD.getWorldTime();
            if (wt) {
                const worldTime = +wt.worldTime;
                const timeRate = +wt.timeRate;

                if (Number.isFinite(worldTime) && worldTime !== this._cache.worldTime) {
                    this._cache.worldTime = worldTime;
                    layer.setText(this._id.worldTime, "WorldTime: " + worldTime);
                }
                if (Number.isFinite(timeRate) && timeRate !== this._cache.timeRate) {
                    this._cache.timeRate = timeRate;
                    layer.setText(this._id.timeRate, "timeRate: " + timeRate);
                }
            }
        }
    }

    /**
     * Destroy whole layer deterministically.
     */
    destroy() {
        if (!this.layer) return;
        this.layer.destroy();
        this.layer = null;
        this._invalidateCache();
    }

    /**
     * Force next update to push all values.
     */
    _invalidateCache() {
        const c = this._cache;
        c.fps = -1;

        c.posXq = 0x7fffffff;
        c.posYq = 0x7fffffff;
        c.posZq = 0x7fffffff;
        c.posStr = "";

        c.camType = "";
        c.camYaw = NaN;
        c.camPitch = NaN;

        c.worldTime = NaN;
        c.timeRate = NaN;
    }
}

module.exports = PlayerUI;

File: assets/Scripts/player/PlayerUI.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaNumber = luaRuntime.LuaNumber
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaArrayIsArray = luaRuntime.LuaArrayIsArray
local LuaClass = luaRuntime.LuaClass
function num(self, v, fb)
    v = LuaNumber(v)
    local lua_Number_isFinite_result_0
    if LuaNumberIsFinite(v) then
        lua_Number_isFinite_result_0 = v
    else
        lua_Number_isFinite_result_0 = fb
    end
    return lua_Number_isFinite_result_0
end
function isObj(self, v)
    return not not v and KTypeOf(v) == "table" and not LuaArrayIsArray(v)
end
function fmt2FromQ(self, q)
    local neg = q < 0
    if neg then
        q = LuaNumber(-q)
    end
    local i = bit32.bor(q / 100, 0)
    local f = q - i * 100
    return ((((neg and "-" or "") .. tostring(i)) .. ".") .. (f < 10 and "0" or "")) .. tostring(f)
end
PlayerUI = LuaClass()
PlayerUI.name = "PlayerUI"
function PlayerUI.prototype.lua_constructor(self, player)
    self.player = player
    local lua_temp_1
    if player and player.cfg and player.cfg.ui then
        lua_temp_1 = player.cfg.ui
    else
        lua_temp_1 = KObject:create(nil)
    end
    local c = lua_temp_1
    self.layerName = tostring(c.layerName or "debug-ui")
    self.anchor = tostring(c.anchor or "tl")
    self.mx = num(_G, c.marginLeft, 10)
    self.my = num(_G, c.marginTop, 10)
    self.w = num(_G, c.w, 280)
    self.padX = num(_G, c.padX, 12)
    self.padY = num(_G, c.padY, 8)
    self.fontTitle = num(_G, c.fontTitle, 16)
    self.fontLine = num(_G, c.fontLine, 14)
    self.gap = num(_G, c.lineGap, 4)
    self.idPrefix = tostring(c.idPrefix or "player.debug")
    local lua_isObj_result_2
    if isObj(_G, c.style) then
        lua_isObj_result_2 = c.style
    else
        lua_isObj_result_2 = {bg = {color = "#0b0f14", alpha = 0.65}, border = {size = 1, color = "#8AA0B6", alpha = 0.45, radius = 8}}
    end
    self.style = lua_isObj_result_2
    self.layer = nil
    self._id = KObject:freeze({
        panel = self.idPrefix .. ".panel",
        title = self.idPrefix .. ".title",
        fps = self.idPrefix .. ".fps",
        pos = self.idPrefix .. ".pos",
        camType = self.idPrefix .. ".camType",
        camYaw = self.idPrefix .. ".camYaw",
        camPitch = self.idPrefix .. ".camPitch",
        worldTime = self.idPrefix .. ".worldTime",
        timeRate = self.idPrefix .. ".timeRate"
    })
    self._cache = {
        fps = -1,
        posXq = 2147483647,
        posYq = 2147483647,
        posZq = 2147483647,
        posStr = "",
        camType = "",
        camYaw = 0 / 0,
        camPitch = 0 / 0,
        worldTime = 0 / 0,
        timeRate = 0 / 0
    }
end
function PlayerUI.prototype.create(self)
    if self.layer then
        return self
    end
    local HUD = self.player and self.player.d and self.player.d.hud
    if not HUD then
        return self
    end
    local layer = HUD:layer(self.layerName)
    self.layer = layer
    layer:spec(
        self:_buildSpec(),
        {relayout = true}
    )
    self:_invalidateCache()
    self:update()
    return self
end
function PlayerUI.prototype._buildSpec(self)
    local id = self._id
    return {
        type = "Panel",
        id = id.panel,
        w = self.w,
        h = 1,
        autoHeight = true,
        padX = self.padX,
        padY = self.padY,
        flow = {fontSize = self.fontLine, gap = self.gap},
        place = {anchor = self.anchor, x = self.mx, y = self.my},
        style = self.style,
        children = {
            {
                type = "Text",
                id = id.title,
                text = "DEBUG",
                fontSize = self.fontTitle,
                color = "#FFFFFF"
            },
            {type = "Text", id = id.fps, text = "FPS: --"},
            {type = "Text", id = id.pos, text = "POS: --"},
            {type = "Text", id = id.camType, text = "CAM(type): --"},
            {type = "Text", id = id.camYaw, text = "CAM(yaw): --"},
            {type = "Text", id = id.camPitch, text = "CAM(pitch): --"},
            {type = "Text", id = id.worldTime, text = "WorldTime: --"},
            {type = "Text", id = id.timeRate, text = "timeRate: --"}
        }
    }
end
function PlayerUI.prototype.refresh(self)
    self:update()
end
function PlayerUI.prototype.update(self, model)
    local layer = self.layer
    if not layer then
        return
    end
    local player = self.player
    local frame = model or player and player.frame
    if not frame then
        return
    end
    local pose = frame.pose
    local view = frame.view
    local eng = player and player.d and player.d.engine
    local lua_temp_3
    if eng and KTypeOf(eng.fps) == "function" then
        lua_temp_3 = bit32.bor(
            eng:fps(),
            0
        )
    else
        lua_temp_3 = 0
    end
    local fps = lua_temp_3
    if fps ~= self._cache.fps then
        self._cache.fps = fps
        layer:setText(
            self._id.fps,
            "FPS: " .. tostring(fps)
        )
    end
    if pose then
        local xq = math.floor(pose.x * 100 + 0.5)
        local yq = math.floor(pose.y * 100 + 0.5)
        local zq = math.floor(pose.z * 100 + 0.5)
        if xq ~= self._cache.posXq or yq ~= self._cache.posYq or zq ~= self._cache.posZq then
            self._cache.posXq = xq
            self._cache.posYq = yq
            self._cache.posZq = zq
            local s = (((("POS: " .. fmt2FromQ(_G, xq)) .. " | ") .. fmt2FromQ(_G, yq)) .. " | ") .. fmt2FromQ(_G, zq)
            self._cache.posStr = s
            layer:setText(self._id.pos, s)
        end
    end
    if view then
        local camType = tostring(view.type)
        if camType ~= self._cache.camType then
            self._cache.camType = camType
            layer:setText(self._id.camType, "CAM(type): " .. camType)
        end
        local yaw = LuaNumber(view.yaw)
        if LuaNumberIsFinite(yaw) and yaw ~= self._cache.camYaw then
            self._cache.camYaw = yaw
            layer:setText(
                self._id.camYaw,
                "CAM(yaw): " .. tostring(yaw)
            )
        end
        local pitch = LuaNumber(view.pitch)
        if LuaNumberIsFinite(pitch) and pitch ~= self._cache.camPitch then
            self._cache.camPitch = pitch
            layer:setText(
                self._id.camPitch,
                "CAM(pitch): " .. tostring(pitch)
            )
        end
    end
    if KTypeOf(ENGINE.world) ~= "nil" and ENGINE.world and KTypeOf(ENGINE.world.getWorldTime) == "function" then
        local wt = ENGINE.world:getWorldTime()
        if wt then
            local worldTime = LuaNumber(wt.worldTime)
            local timeRate = LuaNumber(wt.timeRate)
            if LuaNumberIsFinite(worldTime) and worldTime ~= self._cache.worldTime then
                self._cache.worldTime = worldTime
                layer:setText(
                    self._id.worldTime,
                    "WorldTime: " .. tostring(worldTime)
                )
            end
            if LuaNumberIsFinite(timeRate) and timeRate ~= self._cache.timeRate then
                self._cache.timeRate = timeRate
                layer:setText(
                    self._id.timeRate,
                    "timeRate: " .. tostring(timeRate)
                )
            end
        end
    end
end
function PlayerUI.prototype.destroy(self)
    if not self.layer then
        return
    end
    self.layer:destroy()
    self.layer = nil
    self:_invalidateCache()
end
function PlayerUI.prototype._invalidateCache(self)
    local c = self._cache
    c.fps = -1
    c.posXq = 2147483647
    c.posYq = 2147483647
    c.posZq = 2147483647
    c.posStr = ""
    c.camType = ""
    c.camYaw = 0 / 0
    c.camPitch = 0 / 0
    c.worldTime = 0 / 0
    c.timeRate = 0 / 0
end
M = PlayerUI

return M
